code/datums/components/proximity_monitor.dm

```diff
@@ -1,139 +1,517 @@
+ * * atom/old_loc - the location `moved_atom` just moved from
+ * * dir - the direction `moved_atom` just moved in
  */
-/datum/component/proximity_monitor/proc/HandleMove(datum/source, atom/OldLoc, Dir, forced)
-	if(!Dir) // No dir means the parent teleported, or moved in a non-standard way like getting placed into disposals, onto a table, dropped, picked up, etc.
-		recenter_prox_checkers()
+/datum/component/proximity_monitor/proc/on_nested_loc_move(atom/moved_atom, atom/old_loc, dir)
+	SIGNAL_HANDLER
+
+	// It was just a normal tile-based move, so we return here.
+	if(dir)
+		move_prox_checkers(dir)
+		return
+
+	// Moving onto a turf.
+	if(!isturf(old_loc) && isturf(moved_atom.loc))
+		toggle_checkers(TRUE)
+
+	map_nested_locs()
+	recenter_prox_checkers()
 
 /**
- * Called in Initialize(). Generates a set of `/obj/effect/abstract/proximity_checker` objects around the parent, and registers signals to them.
+ * Called when the receiver or an atom in the `nested_receiver_locs` list moves into a disposals holder object.
+ *
+ * This proc recieves arguments, but they aren't needed.
+ */
+/datum/component/proximity_monitor/proc/on_disposal_enter(datum/source)
+	SIGNAL_HANDLER
+
+	toggle_checkers(FALSE)
+
+/**
+ * Called when the receiver or an atom in the `nested_receiver_locs` list moves out of a disposals holder object.
+ *
+ * This proc recieves arguments, but they aren't needed.
+ */
+/datum/component/proximity_monitor/proc/on_disposal_exit(datum/source)
+	SIGNAL_HANDLER
+
+	toggle_checkers(TRUE)
+
+/**
+ * Registers signals to any nested locations the `hasprox_receiver` is in, excluding turfs, so they can be monitored for movement.
+ */
+/datum/component/proximity_monitor/proc/map_nested_locs()
+	clear_nested_locs()
+	var/atom/loc_to_check = hasprox_receiver.loc
+
+	while(loc_to_check && !isturf(loc_to_check))
+		if(loc_to_check in nested_receiver_locs)
+			continue
+		nested_receiver_locs += loc_to_check
+		RegisterSignal(loc_to_check, COMSIG_MOVABLE_MOVED, .proc/on_nested_loc_move)
+		RegisterSignal(loc_to_check, COMSIG_MOVABLE_DISPOSING, .proc/on_disposal_enter)
+		RegisterSignal(loc_to_check, COMSIG_MOVABLE_EXIT_DISPOSALS, .proc/on_disposal_exit)
+		loc_to_check = loc_to_check.loc
+
+/**
+ * Removes and unregisters signals from all objects currently in the `nested_receiver_locs` list.
+ */
+/datum/component/proximity_monitor/proc/clear_nested_locs()
+	for(var/nested_loc in nested_receiver_locs)
+		UnregisterSignal(nested_loc, list(COMSIG_MOVABLE_MOVED, COMSIG_MOVABLE_DISPOSING, COMSIG_MOVABLE_EXIT_DISPOSALS))
+	nested_receiver_locs = list()
+
+/**
+ * Relays basic directional movement from the `hasprox_receiver` or `host`, to all objects in the `proximity_checkers` list.
+ *
+ * Arguments:
+ * * move_dir - the direction the checkers should move in
+ */
+/datum/component/proximity_monitor/proc/move_prox_checkers(move_dir)
+	for(var/obj/P as anything in proximity_checkers)
+		P.loc = get_step(P, move_dir)
```

Comment:
The proximity checkers are essentially invisible objects which only exist to say to the receiver "hey something moved within X tiles of you". The overhead for `forceMove` is very unnecessary, because turfs and areas don't need to know when these have entered or exited them, etc. Really the only thing that needs to care about what these objects are doing is the component. 

I can add a comment explaining this somewhere in this file.